services/datalad/datalad_service/tasks/description.py

```diff
@@ -11,26 +12,33 @@ def edit_description(description, new_fields):
     return updated
 
 
-def update_description(store, dataset, description_fields, name=None, email=None):
+async def update_description(store, dataset, description_fields, name=None, email=None):
     repo = store.get_dataset_repo(dataset)
-    description = git_show(
+    description_stream, description_size = await git_show_content(
         repo, 'HEAD', 'dataset_description.json')
+    all_content_bytes_list = []
+    async for chunk in description_stream:
+        all_content_bytes_list.append(chunk)
+    description = b"".join(all_content_bytes_list).decode()
     description_json = json.loads(description)
     if description_json.get('License') != 'CC0':
         description_fields = edit_description(
             description_fields, {'License': 'CC0'})
     if description_fields is not None and any(description_fields):
         updated = edit_description(description_json, description_fields)
-        path = os.path.join(store.get_dataset_path(
-            dataset), 'dataset_description.json')
-        with open(path, 'r+', encoding='utf-8', newline='') as description_file:
-            description_file_contents = description_file.read()
+        path = os.path.join(store.get_dataset_path(dataset), 'dataset_description.json')
+        real_path = os.path.realpath(path)
+        # Read the contents (may be annexed)
+        async with aiofiles.open(real_path, 'r+', encoding='utf-8', newline='') as description_file:
+            description_file_contents = await description_file.read()
             if description != description_file_contents:
                 raise Exception('unexpected dataset_description.json contents',
                                 description, description_file_contents)
-            description_file.seek(0)
-            description_file.truncate(0)
-            description_file.write(json.dumps(
+        # If contents are sensible, reopen and overwrite (at the symlink target for annexed files)
+        async with aiofiles.open(path, "a+", encoding='utf-8', newline='') as description_file:
+            await description_file.seek(0)
+            await description_file.truncate(0)
+            await description_file.write(json.dumps(
                 updated, indent=4, ensure_ascii=False))
```

Comment:
I think this will corrupt the file in the annex tree? I'd say it's probably better to unlink and then re-create.

```suggestion
        # If contents are sensible, overwrite (removing the file to avoid 
        await aiofiles.os.unlink(path)
        async with aiofiles.open(path, 'w', encoding='utf-8', newline='') as description_file:
            await description_file.write(json.dumps(
                updated, indent=4, ensure_ascii=False))
```

If we're worried about a crash between these two, you might consider https://boltons.readthedocs.io/en/latest/fileutils.html#atomic-file-saving.

Reply:
Thanks, I added a test to cover a dataset_description.json edit and consolidated the write logic since CHANGES was properly handling this.